RavenDB-19992: Low level optimization of key encoding process

redknightlois opened the pull request:
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-19992

### Additional description
Assembler level optimization to achieve faster encoding of keys. 

### Type of change
- Optimization

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### Is it platform specific issue?
- No

redknightlois committed:
RavenDB-19992: Low level optimization of the encoding process.

redknightlois commented:
![image](https://user-images.githubusercontent.com/1045411/226485345-5bed93ec-d5ed-4c69-afef-471ecb9da1b4.png)

redknightlois committed:
RavenDB-19992: Alternative implementation of the PersistanceDictionaryLocator

After talking with @ayende we agreed that given the idea of the locator is to try to avoid as much the code complexity of having to deal with the collisions with complex codepaths. We decided to settle for a suboptimal solution but continue with it until it becomes an actual performance issue.

ayende commented on src/Voron/Data/CompactTrees/PersistentDictionaryLocator.cs:
```diff
@@ -65,14 +55,33 @@ public void Set(long pageNumber, PersistentDictionary dictionary)
+                        return; // should never happen
+                    _cache[position] = new Dictionary<long, PersistentDictionary>
+                    {
+                        [existing.DictionaryId] = existing,
+                        [dictionary.DictionaryId] = dictionary
```
Subtle behavior here. 
If we have two dictionaries that are on the _same_ id, we'll use the `dictionary` one, not the `existing`, it is not an error in this case.

redknightlois committed:
RavenDB-20142: Potential issue with key state change.

Reusing a compact key during the process of splitting may cause issues if the compact key is not expected to be changed by the caller

redknightlois committed:
RavenDB-20143: Disable the Compact Key pooling. 

It is not clear yet why reusing compact key instances causes corruptions, however, given that the impact of the pooling is reduced, it is preferable to remove temporarily until it can be investigated in full.

redknightlois commented:
@arekpalinski I solved both issues. For RavenDB-20142 I found and solved it. However, for RavenDB-20143 I decided to just get rid of the compact key sharing. With the current usage pattern the contribution of allocations of the compact keys is much smaller than it was initially, therefore, while I didnt remove the infrastructure it is safe to remove this ability until we can investigate deeply why and how sharing compact keys instances corrupt the underlying storage.

arekpalinski commented:
@redknightlois Please check @ayende's comment. I'm merging it and will verify the fix.
